File: frontend/src/types/game-state.ts
// To parse this data:
//
//   import { Convert, GameState } from "./file";
//
//   const gameState = Convert.toGameState(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

/**
 * Complete game state for Carcassonne: War of Ages
 */
export interface GameState {
    /**
     * Remaining tiles in the deck
     */
    availableTiles: AvailableTile[];
    /**
     * ID of the player whose turn it is
     */
    currentPlayer: number;
    /**
     * Three tile options for current player, null if not current player's turn
     */
    currentTileOptions?: TypeElement[] | null;
    /**
     * Recent game events for replay/logging
     */
    events?: Event[];
    /**
     * Unique identifier for the game session
     */
    gameId: string;
    /**
     * Game configuration settings
     */
    gameSettings?: GameSettings;
    /**
     * Timestamp when game started
     */
    gameStartTime: number;
    /**
     * Timestamp of last state update
     */
    lastUpdate: number;
    /**
     * Array of players in the game
     */
    players: Player[];
    /**
     * Current game status
     */
    status: Status;
    /**
     * Array of all tiles placed on the board
     */
    tiles: { [key: string]: any }[];
    /**
     * Current turn number
     */
    turnNumber: number;
    /**
     * Time remaining for current turn in seconds
     */
    turnTimeRemaining: number;
    /**
     * Array of all units on the board
     */
    units: { [key: string]: any }[];
    /**
     * ID of winning player, null if game not finished
     */
    winner?: number | null;
}

export interface AvailableTile {
    count: number;
    type:  TypeElement;
}

export enum TypeElement {
    Barracks = "barracks",
    CapitalCity = "capital_city",
    City = "city",
    Field = "field",
    Marsh = "marsh",
    Mine = "mine",
    Monastery = "monastery",
    Orchard = "orchard",
    Watchtower = "watchtower",
}

export interface Event {
    /**
     * Event-specific data
     */
    data: { [key: string]: any };
    /**
     * Unique event identifier
     */
    id: string;
    /**
     * ID of player who triggered the event
     */
    playerId?: number | null;
    /**
     * When the event occurred
     */
    timestamp: number;
    /**
     * Type of event
     */
    type: EventType;
}

/**
 * Type of event
 */
export enum EventType {
    GameEnded = "game_ended",
    GameStarted = "game_started",
    PlayerEliminated = "player_eliminated",
    TileDestroyed = "tile_destroyed",
    TilePlaced = "tile_placed",
    TurnChanged = "turn_changed",
    UnitAttacked = "unit_attacked",
    UnitCreated = "unit_created",
    UnitKilled = "unit_killed",
    UnitMoved = "unit_moved",
}

/**
 * Game configuration settings
 */
export interface GameSettings {
    /**
     * Size of the square map
     */
    mapSize?: number;
    /**
     * Maximum game duration in seconds
     */
    maxGameDuration?: number;
    /**
     * How often resources are updated in seconds
     */
    resourceUpdateInterval?: number;
    /**
     * Turn duration in seconds
     */
    turnDuration?: number;
}

export interface Player {
    /**
     * Position of capital city, null if eliminated
     */
    capitalCity: CapitalCity | null;
    /**
     * Player color (hex code)
     */
    color: string;
    /**
     * Player ID
     */
    id: number;
    /**
     * Whether player is currently connected
     */
    isConnected: boolean;
    /**
     * Whether player has been eliminated
     */
    isEliminated: boolean;
    /**
     * Player name
     */
    name: string;
    /**
     * Current resources
     */
    resources: Resources;
    /**
     * Player statistics
     */
    stats?: Stats;
    /**
     * Current technology level
     */
    techLevel: TechLevel;
}

export interface CapitalCity {
    x: number;
    y: number;
}

/**
 * Current resources
 */
export interface Resources {
    faith: number;
    food:  number;
    gold:  number;
}

/**
 * Player statistics
 */
export interface Stats {
    tilesDestroyed?: number;
    tilesPlaced?:    number;
    unitsCreated?:   number;
    unitsKilled?:    number;
}

/**
 * Current technology level
 */
export enum TechLevel {
    Duchy = "duchy",
    Kingdom = "kingdom",
    Manor = "manor",
}

/**
 * Current game status
 */
export enum Status {
    Finished = "finished",
    Paused = "paused",
    Playing = "playing",
    Starting = "starting",
    Waiting = "waiting",
}

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
export class Convert {
    public static toGameState(json: string): GameState {
        return cast(JSON.parse(json), r("GameState"));
    }

    public static gameStateToJson(value: GameState): string {
        return JSON.stringify(uncast(value, r("GameState")), null, 2);
    }
}

function invalidValue(typ: any, val: any, key: any, parent: any = ''): never {
    const prettyTyp = prettyTypeName(typ);
    const parentText = parent ? ` on ${parent}` : '';
    const keyText = key ? ` for key "${key}"` : '';
    throw Error(`Invalid value${keyText}${parentText}. Expected ${prettyTyp} but got ${JSON.stringify(val)}`);
}

function prettyTypeName(typ: any): string {
    if (Array.isArray(typ)) {
        if (typ.length === 2 && typ[0] === undefined) {
            return `an optional ${prettyTypeName(typ[1])}`;
        } else {
            return `one of [${typ.map(a => { return prettyTypeName(a); }).join(", ")}]`;
        }
    } else if (typeof typ === "object" && typ.literal !== undefined) {
        return typ.literal;
    } else {
        return typeof typ;
    }
}

function jsonToJSProps(typ: any): any {
    if (typ.jsonToJS === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ: any): any {
    if (typ.jsToJSON === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val: any, typ: any, getProps: any, key: any = '', parent: any = ''): any {
    function transformPrimitive(typ: string, val: any): any {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val, key, parent);
    }

    function transformUnion(typs: any[], val: any): any {
        // val must validate against one typ in typs
        const l = typs.length;
        for (let i = 0; i < l; i++) {
            const typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val, key, parent);
    }

    function transformEnum(cases: string[], val: any): any {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases.map(a => { return l(a); }), val, key, parent);
    }

    function transformArray(typ: any, val: any): any {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue(l("array"), val, key, parent);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformDate(val: any): any {
        if (val === null) {
            return null;
        }
        const d = new Date(val);
        if (isNaN(d.valueOf())) {
            return invalidValue(l("Date"), val, key, parent);
        }
        return d;
    }

    function transformObject(props: { [k: string]: any }, additional: any, val: any): any {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue(l(ref || "object"), val, key, parent);
        }
        const result: any = {};
        Object.getOwnPropertyNames(props).forEach(key => {
            const prop = props[key];
            const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
            result[prop.key] = transform(v, prop.typ, getProps, key, ref);
        });
        Object.getOwnPropertyNames(val).forEach(key => {
            if (!Object.prototype.hasOwnProperty.call(props, key)) {
                result[key] = transform(val[key], additional, getProps, key, ref);
            }
        });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val, key, parent);
    }
    if (typ === false) return invalidValue(typ, val, key, parent);
    let ref: any = undefined;
    while (typeof typ === "object" && typ.ref !== undefined) {
        ref = typ.ref;
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
            : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
            : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
            : invalidValue(typ, val, key, parent);
    }
    // Numbers can be parsed by Date but shouldn't be.
    if (typ === Date && typeof val !== "number") return transformDate(val);
    return transformPrimitive(typ, val);
}

function cast<T>(val: any, typ: any): T {
    return transform(val, typ, jsonToJSProps);
}

function uncast<T>(val: T, typ: any): any {
    return transform(val, typ, jsToJSONProps);
}

function l(typ: any) {
    return { literal: typ };
}

function a(typ: any) {
    return { arrayItems: typ };
}

function u(...typs: any[]) {
    return { unionMembers: typs };
}

function o(props: any[], additional: any) {
    return { props, additional };
}

function m(additional: any) {
    return { props: [], additional };
}

function r(name: string) {
    return { ref: name };
}

const typeMap: any = {
    "GameState": o([
        { json: "availableTiles", js: "availableTiles", typ: a(r("AvailableTile")) },
        { json: "currentPlayer", js: "currentPlayer", typ: 0 },
        { json: "currentTileOptions", js: "currentTileOptions", typ: u(undefined, u(a(r("TypeElement")), null)) },
        { json: "events", js: "events", typ: u(undefined, a(r("Event"))) },
        { json: "gameId", js: "gameId", typ: "" },
        { json: "gameSettings", js: "gameSettings", typ: u(undefined, r("GameSettings")) },
        { json: "gameStartTime", js: "gameStartTime", typ: 3.14 },
        { json: "lastUpdate", js: "lastUpdate", typ: 3.14 },
        { json: "players", js: "players", typ: a(r("Player")) },
        { json: "status", js: "status", typ: r("Status") },
        { json: "tiles", js: "tiles", typ: a(m("any")) },
        { json: "turnNumber", js: "turnNumber", typ: 0 },
        { json: "turnTimeRemaining", js: "turnTimeRemaining", typ: 3.14 },
        { json: "units", js: "units", typ: a(m("any")) },
        { json: "winner", js: "winner", typ: u(undefined, u(0, null)) },
    ], false),
    "AvailableTile": o([
        { json: "count", js: "count", typ: 0 },
        { json: "type", js: "type", typ: r("TypeElement") },
    ], false),
    "Event": o([
        { json: "data", js: "data", typ: m("any") },
        { json: "id", js: "id", typ: "" },
        { json: "playerId", js: "playerId", typ: u(undefined, u(0, null)) },
        { json: "timestamp", js: "timestamp", typ: 3.14 },
        { json: "type", js: "type", typ: r("EventType") },
    ], false),
    "GameSettings": o([
        { json: "mapSize", js: "mapSize", typ: u(undefined, 0) },
        { json: "maxGameDuration", js: "maxGameDuration", typ: u(undefined, 3.14) },
        { json: "resourceUpdateInterval", js: "resourceUpdateInterval", typ: u(undefined, 3.14) },
        { json: "turnDuration", js: "turnDuration", typ: u(undefined, 3.14) },
    ], false),
    "Player": o([
        { json: "capitalCity", js: "capitalCity", typ: u(r("CapitalCity"), null) },
        { json: "color", js: "color", typ: "" },
        { json: "id", js: "id", typ: 0 },
        { json: "isConnected", js: "isConnected", typ: true },
        { json: "isEliminated", js: "isEliminated", typ: true },
        { json: "name", js: "name", typ: "" },
        { json: "resources", js: "resources", typ: r("Resources") },
        { json: "stats", js: "stats", typ: u(undefined, r("Stats")) },
        { json: "techLevel", js: "techLevel", typ: r("TechLevel") },
    ], false),
    "CapitalCity": o([
        { json: "x", js: "x", typ: 0 },
        { json: "y", js: "y", typ: 0 },
    ], false),
    "Resources": o([
        { json: "faith", js: "faith", typ: 0 },
        { json: "food", js: "food", typ: 0 },
        { json: "gold", js: "gold", typ: 0 },
    ], false),
    "Stats": o([
        { json: "tilesDestroyed", js: "tilesDestroyed", typ: u(undefined, 0) },
        { json: "tilesPlaced", js: "tilesPlaced", typ: u(undefined, 0) },
        { json: "unitsCreated", js: "unitsCreated", typ: u(undefined, 0) },
        { json: "unitsKilled", js: "unitsKilled", typ: u(undefined, 0) },
    ], false),
    "TypeElement": [
        "barracks",
        "capital_city",
        "city",
        "field",
        "marsh",
        "mine",
        "monastery",
        "orchard",
        "watchtower",
    ],
    "EventType": [
        "game_ended",
        "game_started",
        "player_eliminated",
        "tile_destroyed",
        "tile_placed",
        "turn_changed",
        "unit_attacked",
        "unit_created",
        "unit_killed",
        "unit_moved",
    ],
    "TechLevel": [
        "duchy",
        "kingdom",
        "manor",
    ],
    "Status": [
        "finished",
        "paused",
        "playing",
        "starting",
        "waiting",
    ],
};

File: backend/src/models/game_state.py
"""
Game state model for Carcassonne: War of Ages.
"""

from typing import Optional, List, Dict, Any
from enum import Enum
from pydantic import BaseModel, Field
from .tile import Tile, TileType
from .unit import Unit, Position
from .follower import Follower
from .tech_tree import TechTree, TechLevel


class GameStatus(str, Enum):
    """Available game statuses."""
    WAITING = "waiting"
    STARTING = "starting"
    PLAYING = "playing"
    PAUSED = "paused"
    FINISHED = "finished"




class PlayerStats(BaseModel):
    """Player statistics."""
    tiles_placed: int = Field(ge=0, default=0)
    units_created: int = Field(ge=0, default=0)
    units_killed: int = Field(ge=0, default=0)
    tiles_destroyed: int = Field(ge=0, default=0)


class Player(BaseModel):
    """A player in the game."""
    id: int = Field(ge=0, description="Player ID")
    name: str = Field(description="Player name")
    color: str = Field(description="Player color (hex code)")
    is_connected: bool = Field(description="Whether player is currently connected")
    is_eliminated: bool = Field(description="Whether player has been eliminated")
    resources: Dict[str, int] = Field(description="Current resources")
    tech_level: TechLevel = Field(default=TechLevel.MANOR, description="Current technology level")
    tech_tree: Optional[TechTree] = Field(default=None, description="Player's tech tree")
    capital_city: Optional[Position] = Field(default=None, description="Position of capital city, null if eliminated")
    stats: Optional[PlayerStats] = Field(default=None, description="Player statistics")
    capital_hp: int = Field(ge=0, default=100, description="Current hit points of player's capital city")
    stored_placements: int = Field(ge=0, le=3, default=0, description="Number of stored tile placements (max 3)")
    followers_available: int = Field(ge=0, le=8, default=8, description="Number of followers in player's pool")

    class Config:
        """Pydantic configuration."""
        use_enum_values = True


class AvailableTile(BaseModel):
    """Available tile type and count."""
    type: TileType = Field(description="Type of tile")
    count: int = Field(ge=0, description="Number of tiles remaining")


class GameEvent(BaseModel):
    """A game event for logging/replay."""
    id: str = Field(description="Unique event identifier")
    type: str = Field(description="Type of event")
    player_id: Optional[int] = Field(ge=0, default=None, description="ID of player who triggered the event")
    timestamp: float = Field(description="When the event occurred")
    data: Dict[str, Any] = Field(description="Event-specific data")


class GameSettings(BaseModel):
    """Game configuration settings."""
    max_players: int = Field(ge=2, le=4, default=4, description="Maximum number of players in the game")
    turn_duration: float = Field(ge=1, default=15.0, description="Turn duration in seconds")
    max_game_duration: float = Field(ge=1, default=1800.0, description="Maximum game duration in seconds")
    map_size: int = Field(ge=10, le=100, default=20, description="Size of the square map")
    resource_update_interval: float = Field(ge=0.1, default=1.0, description="How often resources are updated in seconds")


class GameState(BaseModel):
    """Complete game state for Carcassonne: War of Ages."""
    game_id: str = Field(description="Unique identifier for the game session")
    status: GameStatus = Field(description="Current game status")
    current_player: int = Field(ge=0, description="ID of the player whose turn it is")
    turn_number: int = Field(ge=0, description="Current turn number")
    turn_time_remaining: float = Field(ge=0, description="Time remaining for current turn in seconds")
    game_start_time: float = Field(description="Timestamp when game started")
    last_update: float = Field(description="Timestamp of last state update")
    players: List[Player] = Field(min_length=2, max_length=4, description="Array of players in the game")
    tiles: List[Tile] = Field(description="Array of all tiles placed on the board")
    units: List[Unit] = Field(description="Array of all units on the board")
    followers: List[Follower] = Field(default_factory=list, description="Array of all followers in the game")
    available_tiles: List[AvailableTile] = Field(description="Remaining tiles in the deck")
    current_tile_options: Optional[List[TileType]] = Field(default=None, min_length=3, max_length=3, description="Three tile options for current player")
    winner: Optional[int] = Field(ge=0, default=None, description="ID of winning player, null if game not finished")
    game_settings: Optional[GameSettings] = Field(default=None, description="Game configuration settings")
    events: Optional[List[GameEvent]] = Field(default=None, description="Recent game events for replay/logging")

    class Config:
        """Pydantic configuration."""
        use_enum_values = True
        validate_assignment = True
        schema_extra = {
            "example": {
                "game_id": "game_12345",
                "status": "playing",
                "current_player": 1,
                "turn_number": 10,
                "turn_time_remaining": 12.5,
                "game_start_time": 1640995200.0,
                "last_update": 1640995800.0,
                "players": [
                    {
                        "id": 1,
                        "name": "Alice",
                        "color": "#FF0000",
                        "is_connected": True,
                        "is_eliminated": False,
                        "resources": {
                            "gold": 100,
                            "food": 50,
                            "faith": 25
                        },
                        "tech_level": "manor",
                        "capital_city": {
                            "x": 20,
                            "y": 20
                        },
                        "stats": {
                            "tiles_placed": 5,
                            "units_created": 3,
                            "units_killed": 1,
                            "tiles_destroyed": 0
                        }
                    }
                ],
                "tiles": [],
                "units": [],
                "available_tiles": [
                    {
                        "type": "city",
                        "count": 15
                    }
                ],
                "current_tile_options": ["city", "field", "monastery"],
                "winner": None,
                "game_settings": {
                    "turn_duration": 15.0,
                    "max_game_duration": 1800.0,
                    "map_size": 20,
                    "resource_update_interval": 1.0
                },
                "events": []
            }
        } 